### DISCUSSION TITLE
How to use `LayoutHelper.GetLayoutScale()` For Collapsed or Non-Visible Controls

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#9077] [Q&A]
In the ColorSpectrum, `LayoutHelper.GetLayoutScale(this)` is used to translate between the control's default coordinate system and the displayed coordinate system. The spectrum bitmap itself is rendered taking into account any display scaling factor to avoid color banding issues.

This works fine as long as the ColorSpectrum is in view. However, if you change the Color programmatically or through another tab, the ColorSpectrum will attempt to update it's selection cursor but can't do it correctly because `LayoutHelper.GetLayoutScale(this)` returns 1.0 (the default) as the VisualRoot is null. This method simply doesn't work unless the control is visible and on screen.

What is the proper way of getting the layout scaling factor in this case?

**Comment by maxkatz6** [comment]
I don't think it's possible to get a layout scaling for a control, that isn't attached to any window.
Though my question would be, why do we need to calculate spectrum bitmap data when control isn't attached to the root/window? Can we skip recalculation in this case?

**Reply by robloo** [reply]
> I don't think it's possible to get a layout scaling for a control, that isn't attached to any window.

I'm hoping that isn't true -- perhaps there is another method I can use? Actually, it's not technically the scaling factor I need. It's the raw pixel count information of the display. WinUI has [DisplayInformation](https://learn.microsoft.com/en-us/uwp/api/windows.graphics.display.displayinformation?view=winrt-22621) for this which works everywhere.

In my situation it's probably best just to illustrate. In the `ColorView` it is possible to edit the color in another tab. The `TabControl` swaps out the content in its `ContentPresenter` based on which tab is shown. This means when the palette tab is visible, the spectrum tab contents aren't loaded in the visual tree. However, any color changes to the palette tab should be propagated back to the spectrum selection ellipse. This doesn't currently work:

https://user-images.githubusercontent.com/17993847/193709658-de2dc8dd-d523-4326-bdab-fbdbc7649f34.mp4

> Though my question would be, why do we need to calculate spectrum bitmap data when control isn't attached to the root/window? Can we skip recalculation in this case?

It makes a lot more sense (and simplifies code) to process the changes immediately. All of these primitives live within the same control already. A control should share it's state globally as much as possible.

Of course the simple fix for this is just to update the ellipse position when `OnAttachedToVisualTree` is called within `ColorView`. I can do that but would prefer another method as discussed in the first paragraph.

**Reply by maxkatz6** [reply]
> It's the raw pixel count information of the display. WinUI has DisplayInformation for this which works everywhere

I have used this API before, but don't remember anymore... Wasn't it created depending on the current view/window? As there was "one thread = one window" rule before. So it has information about the window and on which display it is shown. 

Because in Avalonia that's a reason, why GetLayoutScale needs a window information, because depending on the display, there could be a different DPI and pixels count on it.

**Reply by maxkatz6** [reply]
And actually we have similar, but not the same API:
window.PlatformImpl.Screens
I think `window.Screen` or something also exists to get current screen, but I can't check it right now.

**Reply by maxkatz6** [reply]
It still needs a window object though.
For the Color Spectrum another way could be to save last window to get this info, but:
- it might cause menory leaks (as unloaded color picker technically can overlive the window)
- there is a case, when you unload spectrum on one display, move window to a second display, and load spectrum back (by switching tabs)

**Reply by robloo** [reply]
Thanks for your comments.

I decided to just redraw the selection ellipse when re-attached to the visual tree. Your point about moving between screens that may have different display scales would likely require that in the future anyway. I don't think this case is correctly handled in the WinUI version of the control either.

> window.PlatformImpl.Screens

This does look more promising, specifically the [`Screen.PixelDensity`](http://reference.avaloniaui.net/api/Avalonia.Platform/Screen/1C219661) property. As you said though it still requires a window reference. It's cleaner just to refresh in this case.

> For the Color Spectrum another way could be to save last window to get this info, but:

I agree that this is too dangerous and could cause unwanted side-effects.
